### DISCUSSION TITLE
Impossible to use w/o the context manager?

### DISCUSSION THREAD

**Discussion by ZzZombo** [frankie567/httpx-ws#27] [Q&A] [answered]
I need to use it w/o one, and it appears I can't do that. Any solutions?

**Comment by frankie567** [comment, marked as answer]
Yes, it only works as context manager. Under the hood, it uses `httpx.stream`, which is a context manager itself.

If you *really* can't use a context manager, a nice trick is to use an [`ExitStack`](https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack) (or [`AsyncExitStack`](https://docs.python.org/3/library/contextlib.html#contextlib.AsyncExitStack) for async) from the standard `contextlib`.

```py
exit_stack =  contextlib.ExitStack()
ws = exit_stack.enter_context(connect_ws("http://localhost:8000/ws"))
message = ws.receive_text()
exit_stack.close()
```

**Comment by ZzZombo** [comment]
Thank you! I've found just now https://github.com/frankie567/httpx-ws/discussions/11, and have already utilized the hints given there right when you replied. The key limitation was that the entry point is forced to stay synchronous, but I could spin off an asynchronous function in order to create the `httpx` client object previously. However, it all fell apart when I needed to use `httpx-ws` because the context managers would expectedly get torn down on function exit.
